src/coreclr/jit/emit.cpp

```diff
@@ -4479,6 +4610,429 @@ void emitter::emitJumpDistBind()
+        // retain the alignment flag of dstIG->prevIG so the loop
+        // (emitLastInnerLoopStartIgNum ~ emitLastInnerLoopEndIgNum) is still aligned.
+        else if (emitLastInnerLoopStartIgNum != currLoopStart)
+        {
+            // Find the IG before dstIG...
+            instrDescAlign* alignInstr = emitAlignList;
+            while ((alignInstr != nullptr) && (alignInstr->idaIG->igNext != dstIG))
+            {
+                alignInstr = alignInstr->idaNext;
+            }
+
+            // ...and clear the IGF_LOOP_ALIGN flag
+            if (alignInstr != nullptr)
+            {
+                assert(alignInstr->idaIG->igNext == dstIG);
+                alignInstr->idaIG->igFlags &= ~IGF_LOOP_ALIGN;
+            }
+
+            JITDUMP("** Skip alignment for loop IG%02u ~ IG%02u, because it covers an aligned loop IG%02u ~ IG%02u.\n",
+                    currLoopStart, currLoopEnd, emitLastInnerLoopStartIgNum, emitLastInnerLoopEndIgNum);
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------
+//  emitLoopAlignAdjustments: Walk all the align instructions and update them
+//    with actual padding needed.
+
+//  Notes:
+//     For IGs that have align instructions in the end, calculate the actual offset
+//     of loop start and determine how much padding is needed. Based on that, update
+//     the igOffs, igSize and emitTotalCodeSize.
+//
+void emitter::emitLoopAlignAdjustments()
+{
+    // no align instructions
+    if (emitAlignList == nullptr)
+    {
+        return;
+    }
+
+    JITDUMP("*************** In emitLoopAlignAdjustments()\n");
+
+    unsigned short estimatedPaddingNeeded, alignmentBoundary = emitComp->opts.compJitAlignLoopBoundary;
+    unsigned       maxLoopSize = 0;
+
+    if (emitComp->opts.compJitAlignLoopAdaptive)
+    {
+        // For adaptive, adjust the loop size depending on the alignment boundary
+        int maxBlocksAllowedForLoop = genLog2((unsigned)alignmentBoundary) - 1;
+        maxLoopSize                 = alignmentBoundary * maxBlocksAllowedForLoop;
+        estimatedPaddingNeeded      = (alignmentBoundary >> 1) - 1;
+    }
+    else
+    {
+        // For non-adaptive, just take whatever is supplied using COMPlus_ variables
+        maxLoopSize            = emitComp->opts.compJitAlignLoopMaxCodeSize;
+        estimatedPaddingNeeded = alignmentBoundary - 1;
+    }
+
+    unsigned        alignBytesRemoved = 0;
+    unsigned        loopSize          = 0;
+    unsigned        loopIGOffset      = 0;
+    instrDescAlign* alignInstr        = emitAlignList;
+
+    // track the IG that was adjusted so we can update the offsets
+    insGroup* lastIGAdj = emitAlignList->idaIG;
+
+    for (; alignInstr != nullptr; alignInstr = alignInstr->idaNext)
+    {
+        assert(alignInstr->idIns() == INS_align);
+
+        insGroup* alignIG = alignInstr->idaIG;
+
+        // Adjust offsets of all IGs until the current IG
```

Comment:
This confused me for a bit, perhaps a comment explaining the approach would help.

I take it you are updating the offsets of all the IGs between the last IG that had alignment updates an the current IG. None of those blocks have alignment updates so they are not changing size, just offset.

Then you have a fixup loop at the very end to update any IGs that are past the last align IG. 

That way every IG after the first align IG is potentially updated, and only updated once.

Reply:
Since you have a list of all the align instructions, it might be more intuitive to do this update after you process an align IG instead of just before processing the next align IG...?

Then you would not need that final fixup loop as a separate thing.

Reply:
good point. Fixed.